File: akroasis/web/src/audio/autoEqConverter.ts
// Converts AutoEQ parametric profiles to 10-band fixed EQ settings
import type { HeadphoneProfile, ParametricBand } from '../types'

const ISO_FREQUENCIES = [31, 63, 125, 250, 500, 1000, 2000, 4000, 8000, 16000]
const GAIN_MIN = -12
const GAIN_MAX = 12

export function convertToFixedBands(profile: HeadphoneProfile): number[] {
  return ISO_FREQUENCIES.map((centerFreq) => {
    let totalGain = 0
    for (const band of profile.parametricEq) {
      totalGain += calculateGainContribution(band, centerFreq)
    }
    const clamped = Math.max(GAIN_MIN, Math.min(GAIN_MAX, totalGain))
    return Math.round(clamped * 2) / 2
  })
}

function calculateGainContribution(band: ParametricBand, frequency: number): number {
  switch (band.type) {
    case 'peaking':
      return calculatePeakingGain(band, frequency)
    case 'low_shelf':
      return calculateLowShelfGain(band, frequency)
    case 'high_shelf':
      return calculateHighShelfGain(band, frequency)
    case 'low_pass':
    case 'high_pass':
      return 0
  }
}

function calculatePeakingGain(band: ParametricBand, frequency: number): number {
  const w0 = 2 * Math.PI * band.frequency
  const w = 2 * Math.PI * frequency
  const distance = Math.abs(Math.log(w / w0))
  const bandwidth = Math.log(2) / (2 * band.q)

  if (distance < bandwidth * 3) {
    const attenuation = Math.exp(-distance / bandwidth)
    return band.gain * attenuation
  }
  return 0
}

function calculateLowShelfGain(band: ParametricBand, frequency: number): number {
  if (frequency >= band.frequency) {
    return band.gain
  }
  const ratio = frequency / band.frequency
  const slope = Math.max(0.1, Math.min(1.0, band.q))
  const attenuation = Math.max(0, Math.min(1, Math.pow(ratio, slope * 2)))
  return band.gain * attenuation
}

function calculateHighShelfGain(band: ParametricBand, frequency: number): number {
  if (frequency <= band.frequency) {
    return band.gain
  }
  const ratio = band.frequency / frequency
  const slope = Math.max(0.1, Math.min(1.0, band.q))
  const attenuation = Math.max(0, Math.min(1, Math.pow(ratio, slope * 2)))
  return band.gain * attenuation
}

export function searchProfiles(profiles: HeadphoneProfile[], query: string): HeadphoneProfile[] {
  if (!query.trim()) return profiles
  const lower = query.toLowerCase()
  return profiles.filter(
    (p) =>
      `${p.manufacturer} ${p.model}`.toLowerCase().includes(lower) ||
      p.manufacturer.toLowerCase().includes(lower) ||
      p.model.toLowerCase().includes(lower),
  )
}

export function groupByManufacturer(profiles: HeadphoneProfile[]): Map<string, HeadphoneProfile[]> {
  const groups = new Map<string, HeadphoneProfile[]>()
  for (const profile of profiles) {
    const existing = groups.get(profile.manufacturer) ?? []
    existing.push(profile)
    groups.set(profile.manufacturer, existing)
  }
  return groups
}

File: legacy/android/app/src/main/java/app/akroasis/audio/AutoEQConverter.kt
// Converts AutoEQ parametric profiles to fixed-band equalizer settings
package app.akroasis.audio

import app.akroasis.data.model.AutoEQProfile
import app.akroasis.data.model.FilterType
import app.akroasis.data.model.ParametricBand
import kotlin.math.*

object AutoEQConverter {

    fun convertToFixedBands(
        profile: AutoEQProfile,
        centerFrequencies: List<Int>,
        bandLevelRange: ShortArray
    ): List<Short> {
        val minLevel = bandLevelRange[0].toFloat()
        val maxLevel = bandLevelRange[1].toFloat()

        return centerFrequencies.map { centerFreq ->
            var totalGain = 0f

            profile.parametricEq.forEach { band ->
                val gainContribution = calculateGainContribution(band, centerFreq.toFloat())
                totalGain += gainContribution
            }

            val gainInMillibels = (totalGain * 100).coerceIn(minLevel, maxLevel)
            gainInMillibels.toInt().toShort()
        }
    }

    private fun calculateGainContribution(band: ParametricBand, frequency: Float): Float {
        return when (band.type) {
            FilterType.PEAKING -> calculatePeakingGain(band, frequency)
            FilterType.LOW_SHELF -> calculateLowShelfGain(band, frequency)
            FilterType.HIGH_SHELF -> calculateHighShelfGain(band, frequency)
            FilterType.LOW_PASS, FilterType.HIGH_PASS -> 0f
        }
    }

    private fun calculatePeakingGain(band: ParametricBand, frequency: Float): Float {
        val w0 = 2 * PI * band.frequency
        val w = 2 * PI * frequency

        val distance = abs(ln(w / w0).toFloat())
        val bandwidth = ln(2f) / (2 * band.q)

        return if (distance < bandwidth * 3) {
            val attenuation = exp(-distance / bandwidth).toFloat()
            band.gain * attenuation
        } else {
            0f
        }
    }

    private fun calculateLowShelfGain(band: ParametricBand, frequency: Float): Float {
        if (frequency >= band.frequency) {
            return band.gain
        }

        val ratio = frequency / band.frequency
        val slope = band.q.coerceIn(0.1f, 1.0f)
        val attenuation = (ratio.pow(slope * 2)).coerceIn(0f, 1f)

        return band.gain * attenuation
    }

    private fun calculateHighShelfGain(band: ParametricBand, frequency: Float): Float {
        if (frequency <= band.frequency) {
            return band.gain
        }

        val ratio = band.frequency / frequency
        val slope = band.q.coerceIn(0.1f, 1.0f)
        val attenuation = (ratio.pow(slope * 2)).coerceIn(0f, 1f)

        return band.gain * attenuation
    }
}
